plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+                  (case y [ cse1 + (3+x)
+                          , (2+x) + (3+x)
+                          , 4+x
+                          ]
+        ) (1+(2+x))
+
+The second CSE candidate is processed similarly, and the final result is
+
+\x y -> (\cse2 -> (\cse1 -> cse1
+                            +
+                            (case y [ cse1 + (3+x)
+                                    , cse2 + (3+x)
+                                    , 4+x
+                                    ]
+                  ) (1+cse2)
+        ) (2+x)
+
+Here's another example:
+
+force (force ifThenElse
+         (lessThanEqualsInteger 0 0)
+         (delay ((1+2) + (1+2)))
+         (delay (1+2))
+      )
+
+In this case, the first two occurrences of `1+2` can be CSE'd, but the third occurrence
+can not. This is ensured by checking the path when substituting `cse1` for `1+2`. The result is
+
+force (force ifThenElse
+         (lessThanEqualsInteger 0 0)
+         (delay ((\cse1 -> cse1 + cse1) (1+2))
+         (delay (1+2))
+      )
+
+-------------------------------------------------------------------------------
+3. When should CSE run?
+-------------------------------------------------------------------------------
+
+CSE should run for multiple iterations, and should interleave with inlining. The following
+example illustrates why:
+
+\x ->
+  f
+    ((\y -> 1+(y+y)) (0+x))
+    ((\z -> 2+(z+z)) (0+x))
+
+There is no inlining opportunity in this term. After the first iteration of CSE, where
+the common subepxression is `0+x`, we get:
+
+\x ->
+  (\cse1 ->
+    f
+      ((\y -> 1+(y+y)) cse1)
+      ((\z -> 2+(z+z)) cse1)
+  ) (0+x)
+
+Now `y` and `z` can be inlined, after which we get
+
+\x ->
+  (\cse1 ->
+    f
+      (1+(cse1+cse1))
+      (2+(cse1+cse1))
+  ) (0+x)
+
+Now there's a new common subexpression: `cse1+cse1`. So another iteration of CSE is
+needed, yielding:
+
+\x ->
+  (\cse1 ->
+    (\cse2 ->
+      f
+        (1+cse2)
+        (2+cse2)
+    ) (cse1+cse1)
+  ) (0+x)
+
+With this example in mind, one may be tempted to make CSE part of the simplifier, and simply
+run it along with the rest of the simplifier. That is, however, a bad idea. CSE does the reverse
+of inlining; inlining tends to expose more optimization opportunities, and conversely, CSE
+tends to destroy optimization opportunities. Running CSE on a not-fully-optimized program
+may cause many optimization opportunities to be permanently lost.
+
+Therefore, this is what we do: first run the simplifier iterations. Then, run the CSE iterations,
+interleaving with the simplifier. For example, suppose max-simplifier-iterations-uplc=12, and
+max-cse-iterations=4. We first run 12 iterations of the simplifier, then run 4 iterations
+of CSE, with a simplifier pass after each iteration of CSE (i.e., the simplifier is run for a
+total of 16 times).
+
+Finally, since CSE can change the order or the number of occurrences of effects, it is only run
+when conservative optimization is off.
```

Comment:
Can't we make it conservative by only CSE-ing pure expressions?

Reply:
Yeah we can do that later as an improvement.